ios/tunnel/rwcendpoint.go

```diff
@@ -0,0 +1,158 @@
+package tunnel
+
+// I took this code from https://github.com/xjasonlyu/tun2socks which is an amazing project
+// but GPL licensed.
+// Technically this would require me from having go-ios be MIT licensed now.
+// I think though the code is pretty simple and not very original to what tun2socks is about.
+// So I hope it's okay to use this code without changing my license and I can get away with
+// attribution. If not let me know, then I can uhm.. rewrite it.
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/buffer"
+	"gvisor.dev/gvisor/pkg/tcpip"
+	"gvisor.dev/gvisor/pkg/tcpip/header"
+	"gvisor.dev/gvisor/pkg/tcpip/link/channel"
+
+	"gvisor.dev/gvisor/pkg/tcpip/stack"
+)
+
+const (
+	// Queue length for outbound packet, arriving for read. Overflow
+	// causes packet drops.
+	defaultOutQueueLen = 1 << 10
+)
+
+// Endpoint implements the interface of stack.LinkEndpoint from io.ReadWriter.
+type Endpoint struct {
+	*channel.Endpoint
+
+	// rw is the io.ReadWriter for reading and writing packets.
+	rw io.ReadWriter
+
+	// mtu (maximum transmission unit) is the maximum size of a packet.
+	mtu uint32
+
+	// offset can be useful when perform TUN device I/O with TUN_PI enabled.
+	offset int
+
+	// once is used to perform the init action once when attaching.
+	once sync.Once
+
+	// wg keeps track of running goroutines.
+	wg sync.WaitGroup
+}
+
+// New returns stack.LinkEndpoint(.*Endpoint) and error.
+func RWCEndpointNew(rw io.ReadWriter, mtu uint32, offset int) (*Endpoint, error) {
```

Comment:
Is this the endpoint that basically exposes the userspace tunnel?

Reply:
No this is used to connect the virtual TUN device to the lockdown connection readwritecloser

Reply:
I added a comment to clarify.